<a href="https://colab.research.google.com/github/Lord-Kanzler/DS-Unit-1-Sprint-2-Statistics/blob/master/module1/LS_DS13_121_Statistics_Probability_and_Inference.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>

<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Lesson 1*

# Statistics, Probability and Inference


## What is Descriptive Statistics?

<https://statistics.laerd.com/statistical-guides/descriptive-inferential-statistics.php>

In [0]:
import pandas as pd
df = pd.DataFrame({'a': [1,2,3,4,5], 'b': [2,4,6,8,10]})
df.head()

a   b
0  1   2
1  2   4
2  3   6
3  4   8
4  5  10

In [0]:
# How can we quickly look at some descriptive statistics of the above dataframe?
df.describe()

a          b
count  5.000000   5.000000
mean   3.000000   6.000000
std    1.581139   3.162278
min    1.000000   2.000000
25%    2.000000   4.000000
50%    3.000000   6.000000
75%    4.000000   8.000000
max    5.000000  10.000000

## What is Inferential Statistics?

<https://slideplayer.com/slide/5130463/16/images/2/Statistical+Inference.jpg>

## Hypothesis Testing (T-Tests)



### One Sample T-test

Is the sample reflective of our concept of the population?

In [0]:
import numpy as np

coinflips = np.random.binomial(n=1, p=.5, size=12)

coinflips.mean()

0.3333333333333333

What is the population for these coinflips?

The population of all potential coinflips using a "fair" coin. prob=.5



Why via repeated sampling does my sample mean move around a lot?

This is the problem with random sampling --- RANDOMNESS

To fight against the randomness of the sampling process -- We collect more data.

As we collect more data, our conclusions become more precise. Our inferences become more confident.

In [0]:
coinflips = np.random.binomial(n=1, p=.5, size=200)

coinflips.mean()

0.515

The bigger the sample size, the closer the mean gets to .... what?

- True Mean

- Population Mean (population parameter)

**As our sample size goes up, our estimates (of the mean in this case) get more precise, allowing us to infer things about the population with greater confidence.**

### What if I didn't have a fair coin?

Tyler comes up to me and says, for every heads coinflip, Ryan has to give Tyler 2 dollars. For every tails coinflip, Tyler has to give Ryan 1 dollar. 

In [0]:
coinflips = np.random.binomial(n=1, p=.8, size=20)

coinflips

array([0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1])

In [0]:
# Ryan gives Tyler: 
coinflips.sum()*2

30

In [0]:
# Tyler gives Ryan
len(coinflips) - coinflips.sum()

5

If this were a fair coin, I should be paying Tyler double. 

However, based on our small sample, Ryan's paying out 4x more...

Ryan starts to become suspicion about if this is a fair coin or not...

Ryan had in his mind a conception that the population mean for coinflips was .5

However, based on this sample, I"m starting to think that maybe this isn't a fair coin here...

How can I know if it's a fair coin or not?

To really know about this coin, we need a lot of coinflips.

In [0]:
coinflips = np.random.binomial(n=1, p=.8, size=50)

In [0]:
sample_mean = coinflips.mean()

sample_mean

0.88

## T-Test 

Given the data that I have collected. Is the statistic that I have estimated from the sample (mean) **SO** different from my conception of the population, that I would say that the two groups are inherently different?

Are they the same, or are they different (sample vs population)?

1-sample t-test. I have one sample, and I'm comparing to the population.

In [0]:
# TODO - during class, but please help!
from scipy import stats 

Three things to establish before running a T-test

1) Null Hypothesis: (This is a fair coin). The odds should be 50/50. Meaning, **I hypothesize that the sample mean is equal to the population mean.**

$H_0: \mu == .5$

(In the technical notation, a sample mean is always represented as $\bar{x}$)

(In the technical notation, a population mean is always represented as $\mu$)

2) Alternative Hyothesis: This is not a fair coin. I don't know what the probability of heads/tails is, I just hypothesize that it's different from .5. (opposite of my null hypothesis)

$H_a: \mu \neq .5%$

3) Confidence Level:

How certain/confident do I need to be about the situation before I will reject the null hypothesis. 

I want to be **95%** certain (confident) before I will rejec the null hypothesis.

Having a **95%** confidence level means, that I'm willing to allow myself to be tricked by the randomness of the sampling process **5%** of the time.

I'm willing to live with accidentally drawing the wrong conclusion due to randomness 5% of the time.




### Hypothesis Testing

The default state of the world. Usually indicated by my conception of the population. 

In [0]:
# Pass this function my sample, and what I think the population is
# The second value here comes from my hypothesis
# This will run my t-test
stats.ttest_1samp(coinflips, .5)

Ttest_1sampResult(statistic=8.185584150125628, pvalue=9.954547482743331e-11)

4)

Test Statistic: (t-statistic) ~8.18

P-value: .0000000000995

What should my threshold be when evaluating the p-value?

Critical Value: The probability threshold for rejecting vs failure to reject the null hypothesis.

Critical Value = 1-Confidence_Level == .05

P < .05 -> Reject the null hypothesis

P > .05 -> Fail to reject the null hypothesis

5) We need to know what the p-value represents

P-value is the probability of obtaining a sample similar to the one that I collected **if the null hypothesis is true**

If this is a fair coin, there's a .0000000000995 chance that we would have collected. this sample via random sampling.

It's unlikely that it's a fair coin. **It's not certain!!!!**

The p-value is the probability that the null hypothesis is true in the context of the sample that we have collected.

**CONCLUSION: Based on a t-statistic of 8.18, and a p-value of .0000000000995, I reject the null hypothesis that this is a fair coin, and suggest the alternative, that it is an unfair coin. 

I DID NOT SAY THAT I ACCEPT THE ALTERNATIVE HYPOTHESIS


In [0]:
Test

### Load and Clean the Data

In [0]:
# Grab the file from UCI
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data

--2020-02-17 18:18:03--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18171 (18K) [application/x-httpd-php]
Saving to: ‘house-votes-84.data’

house-votes-84.data 100%[===================>]  17.75K  --.-KB/s    in 0.06s   

2020-02-17 18:18:03 (278 KB/s) - ‘house-votes-84.data’ saved [18171/18171]



In [0]:
import pandas as pd

df = pd.read_csv('house-votes-84.data', names=['party','handicapped-infants','water-project',
                          'budget','physician-fee-freeze', 'el-salvador-aid',
                          'religious-groups','anti-satellite-ban',
                          'aid-to-contras','mx-missile','immigration',
                          'synfuels', 'education', 'right-to-sue','crime','duty-free',
                          'south-africa'])

df.head()

party handicapped-infants water-project  ... crime duty-free south-africa
0  republican                   n             y  ...     y         n            y
1  republican                   n             y  ...     y         n            ?
2    democrat                   ?             y  ...     y         n            n
3    democrat                   n             y  ...     n         n            y
4    democrat                   y             y  ...     y         y            y

[5 rows x 17 columns]

In [0]:
df.dtypes

party                   object
handicapped-infants     object
water-project           object
budget                  object
physician-fee-freeze    object
el-salvador-aid         object
religious-groups        object
anti-satellite-ban      object
aid-to-contras          object
mx-missile              object
immigration             object
synfuels                object
education               object
right-to-sue            object
crime                   object
duty-free               object
south-africa            object
dtype: object

In [0]:
df = df.replace({"y": 1, "n": 0, "?": np.NaN})

df.head()

party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

In [0]:
rep = df[df['party']=='republican']

rep.head()

party  handicapped-infants  ...  duty-free  south-africa
0   republican                  0.0  ...        0.0           1.0
1   republican                  0.0  ...        0.0           NaN
7   republican                  0.0  ...        NaN           1.0
8   republican                  0.0  ...        0.0           1.0
10  republican                  0.0  ...        0.0           0.0

[5 rows x 17 columns]

In [0]:
dem = df[df['party']=='democrat']

dem.head()

party  handicapped-infants  water-project  ...  crime  duty-free  south-africa
2  democrat                  NaN            1.0  ...    1.0        0.0           0.0
3  democrat                  0.0            1.0  ...    0.0        0.0           1.0
4  democrat                  1.0            1.0  ...    1.0        1.0           1.0
5  democrat                  0.0            1.0  ...    1.0        1.0           1.0
6  democrat                  0.0            1.0  ...    1.0        1.0           1.0

[5 rows x 17 columns]

In [0]:
df.isnull().sum()

party                     0
handicapped-infants      12
water-project            48
budget                   11
physician-fee-freeze     11
el-salvador-aid          15
religious-groups         11
anti-satellite-ban       14
aid-to-contras           15
mx-missile               22
immigration               7
synfuels                 21
education                31
right-to-sue             25
crime                    17
duty-free                28
south-africa            104
dtype: int64

In [0]:
df['party'].value_counts()

democrat      267
republican    168
Name: party, dtype: int64

### Before Running the Test

On republican support for the handicapped-infants bill

Null Hypothesis: Republicans support this bill 50/50 or that the party is split in their support for this bill. $H_0: \mu == .5$

Alternative Hypothesis: Republican support for this bill is not 50/50. $H_a: \mu \neq .5$

Confidence Level: 95% (My critical value is .05) -> (1-.95 == .05)



### Run the Test

In [0]:
# nan_policy='omit' will automatically drop NaNs
# so that they don't artificially inflate our sample size
# Only considering republican representatives who voted
stats.ttest_1samp(rep['handicapped-infants'], .18, nan_policy='omit')

Ttest_1sampResult(statistic=0.2583045345071252, pvalue=0.7964957641607793)

In [0]:
rep['handicapped-infants'].mean()

0.18787878787878787

### Results and Conclusion

T Statistic: -10.23
P-value: basically 0

Conclusion: 

Based on a t-stat of -10.23 and p-value of 0, I (**reject** / fail to reject) the null hypothesis that republican support for the "handicapped-infants" bill is 50/50. And suggest the alternative, that support is something other than 50%.

p-value is the probability that I would have observed this sample if the null hypothesis were true. -> FAT CHANCE

### Two Sample T-test

Comparing two groups. Are they similar or different?

1) Null Hypothesis: Democrat support for the 'education' bill is equal to republican support for the 'education bill

$ H_0: \bar{x}_{r} == \bar{x}_{d}$

This is the only kind of null hypothesis with a two sample t-test. Are the sample means the same or are they different?

2) Alternative Hypothesis: Democrat support for the 'education' bill is not equal to republican support for the 'education bill

$ H_0: \bar{x}_{r} \neq \bar{x}_{d}$

3) Confidence Level: 95%

In [0]:
stats.ttest_ind(rep['education'], dem['education'], nan_policy='omit')

Ttest_indResult(statistic=20.500685724563073, pvalue=1.8834203990450192e-64)

Conclusion:

Based on a t-statistic of ~20 and a p-value of ~0. I reject the null hypothesis that republicans and democrats support the 'education' bill at the same level/rate. And I suggest the alternative that democrat and republican levels of support for this bill are different. 

In [0]:
rep['education'].mean()

0.8709677419354839

In [0]:
dem['education'].mean()

0.14457831325301204



Ever thought about how long it takes to make a pancake? Have you ever compared the tooking time of a pancake on each eye of your stove? Is the cooking time different between the different eyes? Now, we can run an experiment and collect a sample of 1,000 pancakes on one eye and another 800 pancakes on the other eye. Assumed we used the same pan, batter, and technique on both eyes. Our average cooking times were 180 (5 std) and 178.5 (4.25 std) seconds repsectively. Now, we can tell those numbers are not identicial, but how confident are we that those numbers are practically the same? How do we know the slight difference isn't caused by some external randomness?

Yes, today's lesson will help you figure out how long to cook your pancakes (*theoretically*). Experimentation is up to you; otherwise, you have to accept my data as true. How are going to accomplish this? With probability, statistics, inference and maple syrup (optional). 

<img src="https://images.unsplash.com/photo-1541288097308-7b8e3f58c4c6?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=3300&q=80" width=400>



In [0]:
mu1 = 180 # sample mean
sigma1 = 5 # standard deviation

sample1 = np.random.normal(mu1, sigma1, 1000)

In [0]:
mu2 = 178.5 # sample mean
sigma2 = 4.25 # standard deviation

sample2 = np.random.normal(mu2, sigma2, 800)

As sample sizes increase, our ability to detect small differences is increased. 

Are these to burners inhererntly different in regards to their pancake cooking times



Null H: The two burners cook pancakes at the same rate

Alternative: They cook pancakes at different rates

Confidence Level: 95%

In [0]:
stats.ttest_ind(sample1, sample2)

Ttest_indResult(statistic=5.798654383177605, pvalue=7.879583378678562e-09)

Based on a t-statistic of ~5.8 and a p-value of .00000000788, I reject the null hypothesis, that these two burners cook pancakes at the same rate.

The difference in pancake cooking times between these two burners is not solely due to the randomness of the sampling process, but there is something inherently different about the two burners.

## Understanding the Test Statistic and P-value

https://homepage.divms.uiowa.edu/~mbognar/applets/t.html

![T Statistic Equation](http://lambdachops.com/img/t-statistic-equation.png)

# Resources

- https://homepage.divms.uiowa.edu/~mbognar/applets/t.html
- https://rpsychologist.com/d3/tdist/
- https://gallery.shinyapps.io/tdist/
- https://en.wikipedia.org/wiki/Standard_deviation#Sample_standard_deviation_of_metabolic_rate_of_northern_fulmars
- https://www.khanacademy.org/math/ap-statistics/two-sample-inference/two-sample-t-test-means/v/two-sample-t-test-for-difference-of-means

# Why do we use the t-distribution in hypothesis tests?

![t-distribution-low-dof](https://lambdachops.com/img/t-distribution-low-dof.png)

![t-distribution-approximates-normal](https://github.com/ryanallredblog/ryanallredblog.github.io/blob/master/img/t-distribution-approximates-normal.png?raw=true)

### Helpful video on why we use the t-distribution

<https://www.youtube.com/watch?v=Uv6nGIgZMVw>

However, in order to understand it you'll need to understand what a z-score is:

A z-score calculates the number of standard deviations an observations lies from the population mean. The problem is that in real-world situations, we don't know what the sample mean is, so we have to turn to using the sample mean to estimate the population mean. Because the sample mean is generated from a sample and used to estimate the population mean with some level of uncertainty, it also has its own distribution a nd spread. This means that for low sample sizes both our estimates of the sample mean and sample population are not very precise, they're kind of spread out. It's this spread that makes the t-distribution wider than the normal distribution for low sample sizes. However, with the larger the sample size, the closer the t-distribution approximates the normal distribution.
